components/clientComponents/forms/AddressComplete/AddressComplete.tsx

```diff
@@ -0,0 +1,300 @@
+import { Language } from "@lib/types/form-builder-types";
+import { countries } from "@lib/managedData/countries";
+import { useFeatureFlags } from "@lib/hooks/useFeatureFlags";
+
+interface ManagedComboboxRef {
+  changeInputValue: (value: string) => void;
+}
+
+export const AddressComplete = (props: AddressCompleteProps): React.ReactElement => {
+  const { id, name, required, ariaDescribedBy } = props;
+
+  const [field, meta, helpers] = useField(props);
+
+  const { t, i18n } = useTranslation("form-builder");
+
+  //Address Complete elements
+  const [choices, setChoices] = useState<string[]>([]);
+  const [addressResultCache, setAddressResultCache] = useState<AddressCompleteChoice[]>([]); // Cache the results from the address search.
+
+  // Cache and Allow values
+  const [isReady, setIsReady] = useState(false);
+  const [allow, setAllow] = useState(false);
+  const [apiKey, setApiKey] = useState("");
+
+  const toFullAddress = (address: AddressCompleteChoice): string => {
+    return address.Text + ", " + address.Description;
+  };
+
+  const comboboxRef = useRef<ManagedComboboxRef>(null);
+
+  // Check if addressComplete is allowed.
+  const { getFlag } = useFeatureFlags();
+  const featureFlags = {
+    addressComplete: getFlag("addressComplete"),
+  };
+  useEffect(() => {
+    const checkAllowed = async () => {
+      if (featureFlags.addressComplete) {
+        setAllow(true);
+        setApiKey(process.env.NEXT_PUBLIC_ADDRESSCOMPLETE_API_KEY || "");
+      } else {
+        setApiKey("");
+      }
+
+      setIsReady(true);
+    };
+
+    if (apiKey === "") {
+      // Don't recheck.
+      checkAllowed();
+    }
+  }, [apiKey]);
+
+  //Form fillers address elements
+  const [addressObject, setAddressObject] = useState<AddressElements | null>(
+    field.value ? JSON.parse(field.value) : null
+  );
+
+  // Update the field value when the address object changes
+  useEffect(() => {
+    const newValue = addressObject ? JSON.stringify(addressObject) : "";
+    helpers.setValue(newValue);
+  }, [addressObject, helpers]);
+
+  const onAddressSearch = async (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (!allow) {
+      return;
+    } // Abandon if addressComplete is disabled.
+
+    const query = e.target.value;
+    const responseData = await getAddressCompleteChoices(
+      apiKey,
+      query,
+      addressObject?.country || "CAN"
+    );
+
+    //loop through the responseData and add it to the addressResultsCache
+    const newElements: AddressCompleteChoice[] = [];
+
+    for (let i = 0; i < responseData.length; i++) {
+      // Check key doesn't already exist.
+      if (
+        !addressResultCache.find((item: AddressCompleteChoice) => item.Id === responseData[i].Id)
+      ) {
+        newElements.push(responseData[i]);
+      }
+    }
+
+    if (newElements.length > 0) {
+      setAddressResultCache((prevCache) => [...prevCache, ...newElements]);
+    }
```

Comment:
Seems this cache is not really behaving like a cache- ie, it's not preventing excessive API lookups, as every keypress still sends an API call. 

The cache here seems to have been implemented to address a shortcoming in the original Combobox component where it only accepts display strings with no value or id property, so when selecting an option, in onAddressSet we take the display string and do a lookup in the "cache" to get the id for the retrieval api call. 

Would prefer to fix this in the combobox component by adding an id property that you could then use directly instead of having to do this lookup. 

But maybe that's outside the scope of this PR for the moment.

Reply:
100% out of scope, but yes - that's what it's used for.